### ISSUE TITLE
Escaping curly braces in interpolated strings produces incorrect C# format strings

### ISSUE THREAD

**Issue by swernli** [microsoft/qsharp-compiler#633]
**Describe the bug**

The fix in #469 added handling for escaped open curly brace to interpolated Q# strings, but the results are not correct. Namely, `\{` in a string is replaced with `{` which remains a special character in the generated C# `String.Format` call. C# uses `{{` as the escape in formatted strings, and also requires `}}` for closing braces (the parser doesn't have any explicit handling for closing braces), see [escaping-braces](https://docs.microsoft.com/en-us/dotnet/standard/base-types/composite-formatting#escaping-braces) for more info. We need to decide on how we want to escape curly braces in Q# interpolation, and then produce the right string literal in the QsCompilation for that. Then, if our escape sequence is anything other than C#'s double brace, we'll need to add special handling to the CsharpGeneration step to reinterpret the string and insert `{{` and `}}` where appropriate.

Note that a simple replacement of `\{` with `{{` or even adopting of `{{` as the Q# escape pattern won't always resolve this, as C# has strange behavior when `}}}` is encountered that does not always match intuition (doc link above has an example for this).

**To Reproduce**

```Q#
let r = One;
Message($"\{{r}}");
```

**Expected behavior**

Outputs "{One}"

**Actual behavior**

Outputs "{0}" and the generated C# has `Message.Apply(String.Format("{{0}}", r));`

**Comment by swernli** [comment]
@bettinaheim @alan-geller Here is the issue regarding escaping curly braces in interpolated strings. For what it's worth, the C# docs recommend the (subjectively horrible) approach of using explicit strings for open curly brace and closing curly brace such that they get inserted into the string unambiguously. So something like this works:
```Q#
        let r = One;
        let oCurly = "{";
        let cCurly = "}";
        Message($"{oCurly}{r}{cCurly}");
```
But it does not look very nice.

**Comment by alan-geller** [comment]
> @bettinaheim @alan-geller Here is the issue regarding escaping curly braces in interpolated strings. For what it's worth, the C# docs recommend the (subjectively horrible) approach of using explicit strings for open curly brace and closing curly brace such that they get inserted into the string unambiguously. So something like this works:
> 
> ```qsharp
>         let r = One;
>         let oCurly = "{";
>         let cCurly = "}";
>         Message($"{oCurly}{r}{cCurly}");
> ```
> 
> But it does not look very nice.

That recommendation comes from https://docs.microsoft.com/en-us/dotnet/api/system.string.format?view=netcore-3.1#how-do-i-include-literal-braces--and--in-the-result-string. It is, indeed, pretty ugly, especially since for an interpolated string you need to define the extra symbols.

As that link and others mention, you can use {{ and }} to indicate an escaped, literal { and }, but these are sometimes subject to misinterpretation; $"{{{x}}}" needs to pair the first two { but the last two }.

**Comment by bettinaheim** [comment]
@swernli @alan-geller Aside from any questions on how the generated C# code should look like, I suggest the following behavior for Q#: 
- we don't adopt the C# way of expressing curlies in interpolated strings, but instead allow to use them with backslashes like other special characters
- the backslash would be required for both { and } to use them within interpolated strings, even if they don't match, meaning a spare } in an interpolated string not preceded by backslash would be a Q# compilation error

Do you see any issues with that, or would that make sense?

**Comment by swernli** [comment]
@bettinaheim Yup, makes sense! The the quirks of C# are left to the C# generation, rather than enshrined in the parser/QsCompilation.

**Comment by bettinaheim** [comment]
@swernli Did we ever make progress on this one? I'll tentatively add this to November, though we may need to triage the issues there.

**Comment by swernli** [comment]
@bettinaheim I haven't made the code changes, but the suggestion you had above about how to handle this seems like the right behavior to me. This should definitely be doable for the November timeframe.
